packages/plugin-goat/src/actions.ts

```diff
@@ -0,0 +1,186 @@
+            options?: Record<string, unknown>,
+            callback?: HandlerCallback
+        ): Promise<boolean> => {
+            try {
+                const walletClient = await getWalletClient(runtime);
+                let currentState =
+                    state ?? (await runtime.composeState(message));
+                currentState =
+                    await runtime.updateRecentMessageState(currentState);
+
+                const parameterContext = composeParameterContext(
+                    tool,
+                    currentState
+                );
+                const parameters = await generateParameters(
+                    runtime,
+                    parameterContext
+                );
+
+                const parsedParameters = tool.parameters.safeParse(parameters);
+                if (!parsedParameters.success) {
+                    callback?.({
+                        text: `Invalid parameters for action ${tool.name}: ${parsedParameters.error.message}`,
+                        content: { error: parsedParameters.error.message },
+                    });
+                    return false;
+                }
+
+                const result = await tool.method(
+                    walletClient,
+                    parsedParameters.data
+                );
+                const responseContext = composeResponseContext(
+                    tool,
+                    result,
+                    currentState
+                );
+                const response = await generateResponse(
+                    runtime,
+                    responseContext
+                );
+
+                callback?.({ text: response, content: result });
+                return true;
+            } catch (error) {
+                const errorMessage =
+                    error instanceof Error ? error.message : String(error);
+                callback?.({
+                    text: `Error executing action ${tool.name}: ${errorMessage}`,
+                    content: { error: errorMessage },
+                });
+                return false;
+            }
+        },
+        examples: [],
+    };
+}
+
+function composeParameterContext<TWalletClient extends WalletClient>(
+    tool: DeferredTool<TWalletClient>,
+    state: State
+): string {
+    const contextTemplate = `Respond with a JSON markdown block containing only the extracted values for action "${
+        tool.name
+    }". Use null for any values that cannot be determined.
+
+Example response:
+\`\`\`json
+${parametersToJsonExample(tool.parameters)}
+\`\`\`
+
+{{recentMessages}}
+
+Given the recent messages, extract the following information for the action "${
+        tool.name
+    }":
+${addParametersToDescription("", tool.parameters)}
+`;
+    return composeContext({ state, template: contextTemplate });
+}
+
+async function generateParameters(
+    runtime: IAgentRuntime,
+    context: string
+): Promise<unknown> {
+    return generateObject({
```

Comment:
I would recommend generateObjectV2 instead as it is much more deterministic

Reply:
Neat! 😄  Just moved to that!! Thanks for the rec 🙏